src/scene/materials/shader-graph-builder.js

```diff
@@ -0,0 +1,128 @@
+     * @param {string} name - name of parameter
+     * @param {object} [arguments] - first argument is name string, then 0-N output node ports or params to connect to new core node inputs - number and port type of arguments must match.
+     * @returns {number} returns the created node id
+     * @example
+     * var node0 = pc.shadergraph.addNode('uv0');
+     * var node1 = pc.shadergraph.addNode('texSample', param0, node0 );
+     * var node2 = pc.shadergraph.addNode('splitVec4', node1 );
+     * var node3 = pc.shadergraph.addNode('mul3', param1, node42 );
+     */
+    addNode: function (name) {
+        var args = arguments;
+        var coreNode = this._nodes[name];
+        var sgIndex = this._graph.addSubGraph(coreNode);
+
+        var sgInputs = coreNode.graphData.ioPorts.filter(function (ioPort) {
+            return ioPort.name.startsWith('IN_');
+        });
+
+        if (sgInputs.length === (args.length - 1)) {
+            for (var argIndex = 0; argIndex < (args.length - 1); argIndex++) {
+                var arg = args[argIndex + 1];
+                if (typeof(arg) === 'number') {
+                    // ret port of a node
+                    var argNodeIndex = arg;
+                    this._graph.connect(argNodeIndex, 'OUT_ret', sgIndex, sgInputs[argIndex].name);
+                } else if (arg.type) {
+                    // ioPort (ioPort)
+                    this._graph.connect(-1, arg.name, sgIndex, sgInputs[argIndex].name);
+                } else {
+                    // specific port of a node
+                    this._graph.connect(arg.node, 'OUT_' + arg.port, sgIndex, sgInputs[argIndex].name);
+                }
+            }
+        } else {
+            console.warn("arguments do not match core node");
+        }
+        return sgIndex;
+    },
+    /**
+     * @private
+     * @function
+     * @name pc.ShaderGraphBuilder#addOutput
+     * @description connect a node port to graph material output
+     * @param {any} arg - node port to output
+     * @param {string} type - type of output
+     * @param {string} name - name of output
+     * @example
+     * // valid outputs are:
+     * pc.shadergraph.output(arg, 'vec3', 'vertOff');
+     * pc.shadergraph.output(arg, 'float', 'dAlpha');
+     * pc.shadergraph.output(arg, 'vec3', 'dAlbedo');
+     * pc.shadergraph.output(arg, 'vec3', 'dNormalMap');
+     * pc.shadergraph.output(arg, 'float', 'dGlossiness');
+     * pc.shadergraph.output(arg, 'vec3', 'dSpecularity');
+     * pc.shadergraph.output(arg, 'vec3', 'dEmission');
+     */
+    addOutput: function (arg, type, name) {
+        // assumes this is only called once per graph output TODO: verify this
+        var ioPort = this._graph.addOutput(type, name);
+
+        if (typeof(arg) === 'number') {
+            // ret port of a node
+            var argNodeIndex = arg;
+            this._graph.connect(argNodeIndex, 'OUT_ret', -1, ioPort.name);

```

Comment:
again - please add validation .connect was successful, and useful error message if not

Reply:
As mentioned above, used privately with the editor, it is the editor that does validation for connections - preventing invalid connections from being made - after MVP we can come up with a system to do it with code (maybe even at code build time?).

Reply:
added validation code to the builder class